### PR TITLE
OTT-495: Accepting content object transparency in orb request and passing to Prebid server

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
…parency rules

# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: exchange/utils.go [hunk 1]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+		var (
+			rule openrtb_ext.BidderContentTransparency
+			ok   bool
+		)
+
+		rules, defaultRule := getTransparencyRules(requestExt.Prebid.Transparency)
+
+		for i, bidderRequest := range allBidderRequests {
+
+			newContentObject := &openrtb2.Content{}
```
Comment (pm-nilesh-chate) [inline exchange/utils.go:188]: This approach would have less code branching
```
var newContentObject *openrtb2.Content
...
if len(rules) != 0 {
      ...
      if rule.Include {
             newContentObject = *contentObject
      }
      ...
}
...
```

File: exchange/utils.go [hunk 2]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+
+			bidderRequest.BidRequest = copyRequest(bidderRequest.BidRequest, newContentObject, isApp)
+			allBidderRequests[i] = bidderRequest
+		}
+	}
+}
+
+func copyRequest(originalRequest *openrtb2.BidRequest, contentObject *openrtb2.Content, isApp bool) *openrtb2.BidRequest {
+
+	newRequest := *originalRequest
```
Comment (pm-nilesh-chate) [inline exchange/utils.go:217]: Does this need deep copy for any other fields?

File: exchange/utils.go [hunk 3]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+}
+
+func createNewContentObject(newContentObject, contentObject *openrtb2.Content, include bool, keys []string) {
+
+	if !include {
+		*newContentObject = *contentObject
+		for _, key := range keys {
+
+			switch key {
+			case "id":
```
Comment (pm-nilesh-chate) [inline exchange/utils.go:238]: `reflect.TypeOf` can be used to access the struct properties by their tag name.

File: exchange/utils.go [hunk 4]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+			case "ext":
+				newContentObject.Ext = nil
+			}
+		}
+		return
+	}
+
+	for _, key := range keys {
+		switch key {
+		case "id":
```
Comment (pm-nilesh-chate) [inline exchange/utils.go:297]: similar as above, `reflect.TypeOf` can be used to access the struct properties by their tag name.

**Review by pm-nilesh-chate — COMMENTED** [review]
added a few suggestions. Also, do `go fmt ./...`

File: exchange/utils.go [hunk 5]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
 	return
 }
 
+// updateContentObjectForBidder updates the content object for each bidder based on content transparency rules
+func updateContentObjectForBidder(allBidderRequests []BidderRequest, requestExt *openrtb_ext.ExtRequest) {
+
+	if requestExt != nil && requestExt.Prebid.Transparency != nil {
```
Comment (Pubmatic-Dhruv-Sonone) [inline exchange/utils.go:165]: Move this to util_pubmatic.go

File: exchange/utils.go [hunk 6]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+				newContentObject.Data = contentObject.Data
+			}
+		case "ext":
+			newContentObject.Ext = contentObject.Ext
+		}
+	}
+}
+
+//  getTransparencyRules return content object transparency rule for a bidder and default rule
+func getTransparencyRules(transparencyExt *openrtb_ext.TransparencyExt) (map[string]openrtb_ext.BidderContentTransparency, openrtb_ext.BidderContentTransparency) {
```
Comment (Pubmatic-Dhruv-Sonone) [inline exchange/utils.go:365]: Try removing this function

File: exchange/utils.go [hunk 7]
```diff
@@ -157,6 +159,225 @@ func cleanOpenRTBRequests(ctx context.Context,
...
+
+		for i, bidderRequest := range allBidderRequests {
+
+			newContentObject := &openrtb2.Content{}
+
+			if len(rules) == 0 {
+				newContentObject = nil
+			} else {
+
+				if rule, ok = rules[string(bidderRequest.BidderName)]; !ok {
```
Comment (Pubmatic-Dhruv-Sonone) [inline exchange/utils.go:194]: use :=

File: openrtb_ext/request.go [hunk 8]
```diff
@@ -55,6 +55,16 @@ type ExtRequestPrebid struct {
 
 	CurrencyConversions *ExtRequestCurrency `json:"currency,omitempty"`
 	BidderConfigs       []BidderConfig      `json:"bidderconfig,omitempty"`
+	Transparency        *TransparencyExt    `json:"transparency,omitempty"`
+}
+
+type BidderContentTransparency struct {
```
Comment (pm-viral-vala) [inline openrtb_ext/request.go:61]: Please make struct name consistent at both prebid-server and go-common repository.

**Review by pm-viral-vala — COMMENTED** [review]
Please create master branch PR and assign labels and milestone to each PR

**Pull request merged**